In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("./data/marketing_sample_for_trulia_com-real_estate__20190901_20191031__30k_data.csv")

In [3]:
display(df.shape, df.info(), df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30002 entries, 0 to 30001
Data columns (total 30 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Job Title           30002 non-null  object 
 1   Job Description     30002 non-null  object 
 2   Job Type            0 non-null      float64
 3   Categories          0 non-null      float64
 4   Location            30002 non-null  object 
 5   City                30002 non-null  object 
 6   State               30002 non-null  object 
 7   Country             30002 non-null  object 
 8   Zip Code            16252 non-null  object 
 9   Address             0 non-null      float64
 10  Salary From         0 non-null      float64
 11  Salary To           0 non-null      float64
 12  Salary Period       0 non-null      float64
 13  Apply Url           18392 non-null  object 
 14  Apply Email         0 non-null      float64
 15  Employees           0 non-null      float64
 16  Indu

(30002, 30)

None

Job Title  \
0                  Shift Manager   
1     Operations Support Manager   
2  Senior Product Manager - Data   
3     Part-Time Office Concierge   
4    Print & Marketing Associate   

                                     Job Description  Job Type  Categories  \
0  <div id="jobDescriptionText" class="jobsearch-...       NaN         NaN   
1  <div id="jobDescriptionText" class="jobsearch-...       NaN         NaN   
2  <div id="jobDescriptionText" class="jobsearch-...       NaN         NaN   
3  <div id="jobDescriptionText" class="jobsearch-...       NaN         NaN   
4  <div id="jobDescriptionText" class="jobsearch-...       NaN         NaN   

                  Location           City State        Country Zip Code  \
0  Mission Hills, CA 91345  Mission Hills    CA  United States    91345   
1        Atlanta, GA 30342        Atlanta    GA  United States    30342   
2              Chicago, IL        Chicago    IL  United States      NaN   
3               Festus, MO         Festus    MO  United States      NaN   
4   Cedar Rapids, IA 52404   Cedar Rapids    IA  United States    52404   

   Address  ...  Employer Phone  \
0      NaN  ...             NaN   
1      NaN  ...             NaN   
2      NaN  ...             NaN   
3      NaN  ...             NaN   
4      NaN  ...             NaN   

                                       Employer Logo  \
0  https://d2q79iu7y748jz.cloudfront.net/s/_squar...   
1  https://d2q79iu7y748jz.cloudfront.net/s/_logo/...   
2                                                NaN   
3                                                NaN   
4  https://d2q79iu7y748jz.cloudfront.net/s/_logo/...   

                                  Companydescription        Employer Location  \
0  Del Taco is an American quick service restaura...  Mission Hills, CA 91345   
1  Based in Atlanta, FOCUS Brands Inc. is an inno...                      NaN   
2  Vibes Corp. reputation was built and establish...                      NaN   
3                                                NaN                      NaN   
4  Staples is The Worklife Fulfillment Company, h...   Cedar Rapids, IA 52404   

   Employer City  Employer State  Employer Country Employer Zip Code  \
0  Mission Hills              CA     United States             91345   
1            NaN             NaN     United States               NaN   
2            NaN             NaN     United States               NaN   
3            NaN             NaN     United States               NaN   
4   Cedar Rapids              IA     United States             52404   

                            Uniq Id            Crawl Timestamp  
0  511f9a53920f4641d701d51d3589349f  2019-08-24 09:13:18 +0000  
1  4955daf0a3facbe2acb6c429ba394e6d  2019-09-19 08:16:55 +0000  
2  a0e0d12df1571962b785f17f43ceae12  2019-09-18 02:13:10 +0000  
3  56e411fd731f76ac916bf4fb169250e9  2019-10-24 16:39:13 +0000  
4  3fff5c0ad6981bf4bff6260bd5feab63  2019-08-24 22:29:10 +0000  

[5 rows x 30 columns]

Going forward with these columns for the following reasons:
1. the other columns are only missing values or have too many missing values
2. some of the other columns aren't relevant

In [4]:
df = df[['Job Title','Job Description','City','State','Country','Company Name']]

### EDA (Exploratory Data Analysis)

In [5]:
display(df.head(), df.describe(include='object'))

Job Title  \
0                  Shift Manager   
1     Operations Support Manager   
2  Senior Product Manager - Data   
3     Part-Time Office Concierge   
4    Print & Marketing Associate   

                                     Job Description           City State  \
0  <div id="jobDescriptionText" class="jobsearch-...  Mission Hills    CA   
1  <div id="jobDescriptionText" class="jobsearch-...        Atlanta    GA   
2  <div id="jobDescriptionText" class="jobsearch-...        Chicago    IL   
3  <div id="jobDescriptionText" class="jobsearch-...         Festus    MO   
4  <div id="jobDescriptionText" class="jobsearch-...   Cedar Rapids    IA   

         Country                     Company Name  
0  United States                     Del Taco LLC  
1  United States                     FOCUS Brands  
2  United States                            Vibes  
3  United States  Coldwell Banker Real Estate LLC  
4  United States                   Staples Stores

Job Title                                    Job Description  \
count               30002                                              30002   
unique              18077                                              24714   
top     Account Executive  <div id="jobDescriptionText" class="jobsearch-...   
freq                  329                                                209   

            City  State        Country Company Name  
count      30002  30002          30002        30000  
unique      3472     81              1        12740  
top     New York     CA  United States         Lids  
freq        1260   4495          30002          425

From the info above we can see the following:
1. Company Name has 2 missing values 
2. This is for US based jobs, or from US based companies
3. there are 18077 unique job titles and 24714 unique job descriptions, and 12740 company names.
4. The top position companies are hiring for is Account Executive, occuring 329 times
5. Most jobs in this dataset are for New York occurring 4495 times 
6. The company Lids has the most resume's out at 425 times

In [6]:
display(df['State'].unique(), df['City'].unique())

array(['CA', 'GA', 'IL', 'MO', 'IA', 'DC', 'WA', 'NC', 'NY', 'TX', 'ID',
       'FL', 'ME', 'OH', 'VA', 'MD', 'MA', 'NH', 'United States', 'MI',
       'CO', 'NJ', 'OR', 'OK', 'WY', 'AZ', 'CT', 'WI', 'AR', 'PA', 'SC',
       'KY', 'MN', 'IN', 'NV', 'MT', 'AL', 'UT', 'RI', 'KS', 'TN', 'AK',
       'SD', 'NM', 'NE', 'California', 'WV', 'MS', 'VT', 'Texas', 'HI',
       'DE', 'LA', 'New Jersey', 'ND', 'Missouri', 'Georgia', 'Ohio',
       'New York State', 'Colorado', 'Michigan', 'Massachusetts', 'Tamil',
       'Remote', 'North Carolina', 'Home Based', 'Oklahoma', 'Indiana',
       'Florida', 'Sydney NSW', 'Nevada', 'BC', 'Utah', 'Wisconsin',
       'Kansas', 'Connecticut', 'Iowa', 'South Carolina', 'Hong',
       'Virginia', 'Beerse'], dtype=object)

array(['Mission Hills', 'Atlanta', 'Chicago', ..., 'Chadds Ford',
       'Great Bend', 'Harrison'], dtype=object)

In [7]:
df_remote = df[df['State'] == 'Remote']
df_home_based = df[df['State'] == 'Home Based']

display(df_remote.head(), df_home_based.head())
display(df_remote.shape)

Job Title  \
8489      Onboarding Specialist (Remote)   
16306         Enterprise Account Manager   
19029  Senior Director, Enterprise Sales   
20457        Account Executive - US West   

                                         Job Description    City   State  \
8489   <div id="jobDescriptionText" class="jobsearch-...  Remote  Remote   
16306  <div id="jobDescriptionText" class="jobsearch-...  Remote  Remote   
19029  <div id="jobDescriptionText" class="jobsearch-...  Remote  Remote   
20457  <div id="jobDescriptionText" class="jobsearch-...  Remote  Remote   

             Country            Company Name  
8489   United States               Apollo.io  
16306  United States                  Ekahau  
19029  United States  Seismic Software, Inc.  
20457  United States           Liferay, Inc.

Job Title  \
11683  Regional Sales Director   

                                         Job Description        City  \
11683  <div id="jobDescriptionText" class="jobsearch-...  Home Based   

            State        Country  Company Name  
11683  Home Based  United States  AMETEK, Inc.

(4, 6)

State is either written in letters or full out, we need to ensure consistent naming

In [8]:
df_state_us = df[df['State'] == 'United States']
display(df_state_us.head(), df_state_us.shape)

Job Title  \
27    Account Executive - Acumatica Cloud ERP US Bas...   
416                        Director of Enterprise Sales   
707                                      Sales Director   
1255                               Cloud Data Architect   
1766  Account Manager, Produce - Jackson, MS to Memp...   

                                        Job Description           City  \
27    <div id="jobDescriptionText" class="jobsearch-...  United States   
416   <div id="jobDescriptionText" class="jobsearch-...  United States   
707   <div id="jobDescriptionText" class="jobsearch-...  United States   
1255  <div id="jobDescriptionText" class="jobsearch-...  United States   
1766  <div id="jobDescriptionText" class="jobsearch-...  United States   

              State        Country         Company Name  
27    United States  United States          Net at Work  
416   United States  United States         Voyager Labs  
707   United States  United States  SecurityPoint Media  
1255  United States  United States              Pythian  
1766  United States  United States                 UNFI

(58, 6)

In [9]:
state_dict = {'CA': 'California','GA': 'Georgia', 'IL': 'Illinois', 'MO': 'Missouri', 'IA': 'Iowa', 
               'DC': 'District of Columbia', 'WA': 'Washington', 'NC': 'North Carolina', 'NY': 'New York', 'TX': 'Texas',
               'ID': 'Idaho', 'FL': 'Florida', 'ME': 'Maine', 'OH': 'Ohio', 'VA': 'Virginia', 
               'MD': 'Maryland', 'MA': 'Massachusetts', 'NH': 'New Hampshire', 'MI': 'Michigan', 'CO': 'Colorado', 
               'NJ': 'New Jersey', 'OR': 'Oregon', 'OK': 'Oklahoma', 'WY': 'Wyoming', 'AZ': 'Arizona',
               'CT': 'Connecticut', 'WI': 'Wisconsin', 'AR': 'Arkansas', 'PA': 'Pennsylvania', 'SC': 'South Carolina',
               'KY': 'Kentucky', 'MN': 'Minnesota', 'IN': 'Indiana', 'NV': 'Nevada', 'MT': 'Montana',
               'AL': 'Alabama', 'UT': 'Utah', 'RI': 'Rhode Island', 'KS': 'Kansas', 'TN': 'Tennessee',
               'AK': 'Alaska', 'SD': 'South Dakota', 'NM': 'New Mexico', 'NE': 'Nebraska', 'WV': 'West Virginia',
               'MS': 'Mississippi', 'VT': 'Vermont', 'HI': 'Hawaii', 'DE': 'Delaware', 'LA': 'Louisiana', 'ND': 'North Dakota',
               'Home Based': 'Remote',
               'New York State': 'New York'
    }

In [10]:
def convert_state_names(state):
    return state_dict.get(state, state)

df['State'] = df['State'].apply(convert_state_names)
df.head()


Job Title  \
0                  Shift Manager   
1     Operations Support Manager   
2  Senior Product Manager - Data   
3     Part-Time Office Concierge   
4    Print & Marketing Associate   

                                     Job Description           City  \
0  <div id="jobDescriptionText" class="jobsearch-...  Mission Hills   
1  <div id="jobDescriptionText" class="jobsearch-...        Atlanta   
2  <div id="jobDescriptionText" class="jobsearch-...        Chicago   
3  <div id="jobDescriptionText" class="jobsearch-...         Festus   
4  <div id="jobDescriptionText" class="jobsearch-...   Cedar Rapids   

        State        Country                     Company Name  
0  California  United States                     Del Taco LLC  
1     Georgia  United States                     FOCUS Brands  
2    Illinois  United States                            Vibes  
3    Missouri  United States  Coldwell Banker Real Estate LLC  
4        Iowa  United States                   Staples Stores

In [11]:
display(df['State'].unique(), df['City'].unique())

array(['California', 'Georgia', 'Illinois', 'Missouri', 'Iowa',
       'District of Columbia', 'Washington', 'North Carolina', 'New York',
       'Texas', 'Idaho', 'Florida', 'Maine', 'Ohio', 'Virginia',
       'Maryland', 'Massachusetts', 'New Hampshire', 'United States',
       'Michigan', 'Colorado', 'New Jersey', 'Oregon', 'Oklahoma',
       'Wyoming', 'Arizona', 'Connecticut', 'Wisconsin', 'Arkansas',
       'Pennsylvania', 'South Carolina', 'Kentucky', 'Minnesota',
       'Indiana', 'Nevada', 'Montana', 'Alabama', 'Utah', 'Rhode Island',
       'Kansas', 'Tennessee', 'Alaska', 'South Dakota', 'New Mexico',
       'Nebraska', 'West Virginia', 'Mississippi', 'Vermont', 'Hawaii',
       'Delaware', 'Louisiana', 'North Dakota', 'Tamil', 'Remote',
       'Sydney NSW', 'BC', 'Hong', 'Beerse'], dtype=object)

array(['Mission Hills', 'Atlanta', 'Chicago', ..., 'Chadds Ford',
       'Great Bend', 'Harrison'], dtype=object)

In [12]:
display(df[df['State'] == 'Sydney NSW'], df[df['State'] == 'BC'], df[df['State'] == 'Hong'], df[df['State'] == 'Tamil'], df[df['State'] == 'Beerse'])

Job Title  \
13924  Commercial Solutions Engineer   

                                         Job Description        City  \
13924  <div id="jobDescriptionText" class="jobsearch-...  Sydney NSW   

            State        Country Company Name  
13924  Sydney NSW  United States       VMware

Job Title                                    Job Description  \
18141  Income Auditor  <div id="jobDescriptionText" class="jobsearch-...   

                       City State        Country                 Company Name  
18141  Harrison Hot Springs    BC  United States  Harrison Hot Springs Resort

Job Title  \
27049  Cloud Solution Architect   

                                         Job Description       City State  \
27049  <div id="jobDescriptionText" class="jobsearch-...  Cyberport  Hong   

             Country Company Name  
27049  United States    Microsoft

Job Title  \
7785  Manager, Professional Services - Delivery   

                                        Job Description     City  State  \
7785  <div id="jobDescriptionText" class="jobsearch-...  Chennai  Tamil   

            Country Company Name  
7785  United States       VMware

Job Title  \
27769  Manager, Global Brand Protection Services [Ent...   

                                         Job Description    City   State  \
27769  <div id="jobDescriptionText" class="jobsearch-...  Beerse  Beerse   

             Country                           Company Name  
27769  United States  Johnson & Johnson Family of Companies

Here we some inconsistencies, the country is US but the states are form different countries. we will drop these 

In [13]:
# List of states to be dropped
states_to_drop = ['Sydney NSW', 'BC', 'Hong', 'Tamil', 'Beerse']

# Drop rows where the 'State' column matches any of the states to be dropped
df = df[~df['State'].isin(states_to_drop)]

df.head()


Job Title  \
0                  Shift Manager   
1     Operations Support Manager   
2  Senior Product Manager - Data   
3     Part-Time Office Concierge   
4    Print & Marketing Associate   

                                     Job Description           City  \
0  <div id="jobDescriptionText" class="jobsearch-...  Mission Hills   
1  <div id="jobDescriptionText" class="jobsearch-...        Atlanta   
2  <div id="jobDescriptionText" class="jobsearch-...        Chicago   
3  <div id="jobDescriptionText" class="jobsearch-...         Festus   
4  <div id="jobDescriptionText" class="jobsearch-...   Cedar Rapids   

        State        Country                     Company Name  
0  California  United States                     Del Taco LLC  
1     Georgia  United States                     FOCUS Brands  
2    Illinois  United States                            Vibes  
3    Missouri  United States  Coldwell Banker Real Estate LLC  
4        Iowa  United States                   Staples Stores

In [14]:
display(df['State'].unique(), df['City'].unique())

array(['California', 'Georgia', 'Illinois', 'Missouri', 'Iowa',
       'District of Columbia', 'Washington', 'North Carolina', 'New York',
       'Texas', 'Idaho', 'Florida', 'Maine', 'Ohio', 'Virginia',
       'Maryland', 'Massachusetts', 'New Hampshire', 'United States',
       'Michigan', 'Colorado', 'New Jersey', 'Oregon', 'Oklahoma',
       'Wyoming', 'Arizona', 'Connecticut', 'Wisconsin', 'Arkansas',
       'Pennsylvania', 'South Carolina', 'Kentucky', 'Minnesota',
       'Indiana', 'Nevada', 'Montana', 'Alabama', 'Utah', 'Rhode Island',
       'Kansas', 'Tennessee', 'Alaska', 'South Dakota', 'New Mexico',
       'Nebraska', 'West Virginia', 'Mississippi', 'Vermont', 'Hawaii',
       'Delaware', 'Louisiana', 'North Dakota', 'Remote'], dtype=object)

array(['Mission Hills', 'Atlanta', 'Chicago', ..., 'Chadds Ford',
       'Great Bend', 'Harrison'], dtype=object)

In [15]:
df['Job Description'].duplicated().sum()

5288

In [16]:
duplicate_rows = df[df.duplicated()]
print('Duplicate Rows:')
duplicate_rows

Duplicate Rows:


Job Title  \
1079                               Jr. Account Executive   
1214                                    Leasing Director   
1705                                Sales Representative   
1720   Account Executive - Hunter - IT Managed Servic...   
2459                                  Lids Store Manager   
...                                                  ...   
29900                            Sales Account Executive   
29903                           Mixed Martial Arts Coach   
29963                  Advertising Sales Representatiive   
29977                          Sales and Marketing Agent   
29983             Account Executive, IT Sales Experience   

                                         Job Description           City  \
1079   <div id="jobDescriptionText" class="jobsearch-...      Charlotte   
1214   <div id="jobDescriptionText" class="jobsearch-...     Lake Orion   
1705   <div id="jobDescriptionText" class="jobsearch-...      Baltimore   
1720   <div id="jobDescriptionText" class="jobsearch-...        Houston   
2459   <div id="jobDescriptionText" class="jobsearch-...     Grove City   
...                                                  ...            ...   
29900  <div id="jobDescriptionText" class="jobsearch-...  Overland Park   
29903  <div id="jobDescriptionText" class="jobsearch-...        Concord   
29963  <div id="jobDescriptionText" class="jobsearch-...       Lakewood   
29977  <div id="jobDescriptionText" class="jobsearch-...         Dallas   
29983  <div id="jobDescriptionText" class="jobsearch-...       Holbrook   

                State        Country                           Company Name  
1079   North Carolina  United States  American Financial Education Alliance  
1214         Michigan  United States        Pomeroy Living - Orion Assisted  
1705         Maryland  United States                              Datastory  
1720            Texas  United States                     SEAMLESS SOLUTIONS  
2459     Pennsylvania  United States                                   Lids  
...               ...            ...                                    ...  
29900          Kansas  United States                 Charter Communications  
29903      California  United States                                UFC Gym  
29963        Colorado  United States                           N-Compass TV  
29977           Texas  United States                     1800 Radiator & AC  
29983        New York  United States           Future Tech Enterprise, Inc.  

[1248 rows x 6 columns]

The are 1248 duplicate rows, but 5288 duplicate job descriptions, the discrepency can mean a few things:
1. **Common Job Roles**: Some Job roles might have the same description (further analysis required)
2. **Recruitment Agencies or Job Board**: These companies might post the same jobs on multiple platforms:
3. **Franchise or Branches**: Companies wih multiple franchises or branches may use the same job description across different locations 
4. **Template Usage**: Some companies may use templates, that shares common language

In [17]:
common_job_roles = df['Job Title'].value_counts().head(10)
print('Top 10 most common job roles:')
print(common_job_roles)

Top 10 most common job roles:
Job Title
Account Executive                       329
Sales Associate/Beauty Advisor          209
Account Manager                         192
General Manager                         180
Sales Associate, Retail Part Time       148
Lids Assistant Manager Full-Time        128
Customer Service Representative         124
Store Manager                           120
Lids Store Manager                      119
Lids District Mgr in Training (DMIT)    112
Name: count, dtype: int64


In [18]:
companies_with_duplicates = df[df['Job Description'].duplicated()]['Company Name'].value_counts()
print("Companies with duplicate job descriptions:")
print(companies_with_duplicates)

Companies with duplicate job descriptions:
Company Name
Lids                               415
Sally Beauty Supply LLC            208
GNC - General Nutrition Centers    201
WALGREENS                          196
Jackson Hewitt                     149
                                  ... 
Fleetcor                             1
Mytech Partners                      1
New Seasons Market                   1
Crunch Fitness - Westland            1
N-Compass TV                         1
Name: count, Length: 1157, dtype: int64


In [19]:
duplicate_locations = df[df['Job Description'].duplicated()]['City'].value_counts()
print("Cities with duplicate job descriptions:")
print(duplicate_locations)

Cities with duplicate job descriptions:
City
New York          146
Houston            76
Washington         67
Los Angeles        65
Chicago            56
                 ... 
Northborough        1
Tullahoma           1
Roanoke Rapids      1
Covina              1
Mentor              1
Name: count, Length: 1715, dtype: int64


In [20]:
duplicate_job_desc = df[df['Job Description'].duplicated()]['Job Description']
print("Sample duplicate job descriptions:")
print(duplicate_job_desc.head())


Sample duplicate job descriptions:
17     <div id="jobDescriptionText" class="jobsearch-...
45     <div id="jobDescriptionText" class="jobsearch-...
94     <div id="jobDescriptionText" class="jobsearch-...
97     <div id="jobDescriptionText" class="jobsearch-...
111    <div id="jobDescriptionText" class="jobsearch-...
Name: Job Description, dtype: object


In [21]:
df_no_dupes = df.drop_duplicates() 
display(df_no_dupes.head(), df_no_dupes.shape)

Job Title  \
0                  Shift Manager   
1     Operations Support Manager   
2  Senior Product Manager - Data   
3     Part-Time Office Concierge   
4    Print & Marketing Associate   

                                     Job Description           City  \
0  <div id="jobDescriptionText" class="jobsearch-...  Mission Hills   
1  <div id="jobDescriptionText" class="jobsearch-...        Atlanta   
2  <div id="jobDescriptionText" class="jobsearch-...        Chicago   
3  <div id="jobDescriptionText" class="jobsearch-...         Festus   
4  <div id="jobDescriptionText" class="jobsearch-...   Cedar Rapids   

        State        Country                     Company Name  
0  California  United States                     Del Taco LLC  
1     Georgia  United States                     FOCUS Brands  
2    Illinois  United States                            Vibes  
3    Missouri  United States  Coldwell Banker Real Estate LLC  
4        Iowa  United States                   Staples Stores

(28749, 6)

In [22]:
popular_jobs_per_state = df_no_dupes.groupby('State')['Job Title'].apply(lambda x: x.value_counts().idxmax())
popular_jobs_df = popular_jobs_per_state.reset_index()
popular_jobs_df.columns = ['State', 'Job Title']
popular_jobs_df = popular_jobs_df.sort_values(by='State').reset_index(drop=True)
print("Most popular job titles per state:")
print(popular_jobs_df)


Most popular job titles per state:
                   State                                          Job Title
0                Alabama                                  Account Executive
1                 Alaska                  Sales Associate, Retail Part Time
2                Arizona                              Retail Sales Teammate
3               Arkansas                     Sales Associate/Beauty Advisor
4             California                                  Account Executive
5               Colorado                                    Account Manager
6            Connecticut                              Retail Sales Teammate
7               Delaware                              Retail Sales Teammate
8   District of Columbia                       Business Development Manager
9                Florida                     Sales Associate/Beauty Advisor
10               Georgia                                    General Manager
11                Hawaii                             

### Data Scientist jobs

In [31]:
keywords = ['data science', 'data scientist', 
            'ml engineer', 'machine learning engineer', 
            'machine learning', 
            'data analyst', 
            'business analyst',
            'data engineer',
            'data Architect']

pattern = '|'.join(keywords)

data_jobs = df_no_dupes[df_no_dupes['Job Title'].str.contains(pattern, case=False, regex=True)]

display(data_jobs.head(), data_jobs.shape)

Job Title  \
18   Business Analyst with Adobe Experience Manager...   
36                    Business Analyst/Project Manager   
69   Senior Data Scientist, Wholesale Credit Techno...   
87          Senior Data Engineer- Platform Development   
101  Data Architect - Data Warehouse & MPP - Nation...   

                                       Job Description              City  \
18   <div id="jobDescriptionText" class="jobsearch-...          New York   
36   <div id="jobDescriptionText" class="jobsearch-...  Farmington Hills   
69   <div id="jobDescriptionText" class="jobsearch-...         Charlotte   
87   <div id="jobDescriptionText" class="jobsearch-...            Irvine   
101  <div id="jobDescriptionText" class="jobsearch-...            Dallas   

              State        Country               Company Name  
18         New York  United States                  Capgemini  
36         Michigan  United States              FutureSoft IT  
69   North Carolina  United States            Bank of America  
87       California  United States                   Weedmaps  
101           Texas  United States  Amazon Web Services, Inc.

(373, 6)

In [30]:
from IPython.display import display, HTML 
display(HTML(data_jobs['Job Description'][1041]))

### Data Scraping with beatutifulsoup

In [ ]:
data_jobs['Bullets'] = [[bullet.text.strip() for bullet in soup.find_all('li')] for soup in soup_objects]

In [32]:
bulleted_post_count = 0 
for bullet_list in data_jobs.Bullets:
  if bullet_list:
    bulleted_post_count += 1 
    
percent_bulleted = 100 * bulleted_post_count / data_jobs.shape[0]
print(f"{percent_bulleted:.2f}% of the postings contain bullets")

AttributeError: 'DataFrame' object has no attribute 'Bullets'

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer 

def rank_words(text_list):
  vectorizer = TfidfVectorizer(stop_words='english')
  tfidf_matrix = vectorizer.fit_transform(text_list).toarray() 
  df = pd.DataFrame({'Words': vectorizer.get_feature_names_out(), 'Summed TFIDF': tfidf_matrix.sum(axis=0)})
  sorted_df = df.sort_values('Summed TFIDF', ascending=False)
  return sorted_df

all_bullets = [] 
for bullet_list in data_jobs.Bullets: 
  all_bullets.extend(bullet_list)
  
sorted_df = rank_words(all_bullets)
print(sorted_df[:5].to_string(index=False))

     Words  Summed TFIDF
      data     12.740915
experience     12.292084
  learning      7.958197
  business      7.440381
   machine      7.201028


In [27]:
data_jobs['Job Title'].unique()


array(['Business Analyst with Adobe Experience Manager (AEM) experience',
       'Business Analyst/Project Manager',
       'Senior Data Scientist, Wholesale Credit Technology',
       'Data Strategy Architect - Cloud, Big Data, Analytics, ML/AI/Data Science',
       'Business Analyst & Systems Manager', 'Data Analyst',
       'Jr. Data Analyst Compliance', 'Data Scientist',
       'Customer Data Analyst', 'Business and Data Analyst Manager',
       '(Temporary) Researcher and Data Analyst, Fine Arts',
       'Sr Data Analyst', 'HRIS Business Analyst (Reporting & Analytics)',
       'Senior Financial Reporting & Business Analyst',
       'Sr. Fraud Data Analyst',
       'Technology Project Manager/Business Analyst',
       'Business Analyst / Project Manager / Technical Solutions',
       'Process Engineer / Business Analyst',
       'Business Analyst - Point-Of-Sale -BBB - 494-1',
       'Business Analyst', 'Business Analyst - SaaS',
       'Sr Demand Forecasting Data Scientist',
    